flang/lib/Lower/ConvertVariable.cpp

```diff
@@ -393,6 +406,47 @@ static mlir::Value createNewLocal(Fortran::lower::AbstractConverter &converter,
   return builder.allocateLocal(loc, ty, nm, symNm, shape, lenParams, isTarg);
 }
 
+/// Must \p var be default initialized at runtime when entering its scope.
+static bool
+mustBeDefaultInitializedAtRuntime(const Fortran::lower::pft::Variable &var) {
+  if (!var.hasSymbol())
+    return false;
+  const auto &sym = var.getSymbol();
+  // Local variable (including function results), and intent(out) dummies must
+  // be default initialized at runtime if their type has default initialization.
+  if ((!var.isGlobal() && !Fortran::semantics::IsDummy(sym)) ||
```

Comment:
This negative logic might be improved by writing this as
```c++
if (var.isGlobal())
  return false;
if (...::IsDummy(sym) && ! ...::IsIntentOut(sym))
  return false;
return hasDefaultInit(sym);
```
I think this logic will return `true` for the (buggy) case of both global and intent out returning `true`.

Reply:
Yes, your logic is safer, I took it.